src/calc/src/expr_manager.rs

```diff
@@ -0,0 +1,540 @@
+        // Shunting Yard algorithm. Based on pseudo-code
+        // from [wiki](https://en.wikipedia.org/wiki/Shunting_yard_algorithm).
+
+        // The token queue.
+        let mut postfix: Vec<&Token> = Vec::new();
+        // The option stack for operation tokens.
+        let mut opt_stack: Vec<&Token> = Vec::new();
+
+        for token in tokens {
+            match token.btn {
+                Btn::Num(_) | Btn::Comma | Btn::Const(_) => postfix.push(token),
+                Btn::BinOpt(_) => {
+                    while opt_stack.last().is_some() {
+                        let top = opt_stack.last().unwrap();
+
+                        if top.btn == Btn::BracketLeft
+                            || top.item.priority < token.item.priority
+                            || (top.item.priority == token.item.priority && !token.item.left_asoc)
+                        {
+                            break;
+                        }
+
+                        postfix.push(opt_stack.pop().unwrap());
+                    }
+                    opt_stack.push(token);
+                }
+                Btn::UnaryOpt(_) | Btn::BracketLeft => opt_stack.push(token),
+                Btn::BracketRight => {
+                    // Pop all operators until the left bracket from the operator stack.
+                    while opt_stack.last().is_some()
+                        && opt_stack.last().unwrap().btn != Btn::BracketLeft
+                    {
+                        postfix.push(opt_stack.pop().unwrap());
+                    }
+                    // Check if left bracket was found.
+                    if opt_stack.last().is_none()
+                        || opt_stack.last().unwrap().btn != Btn::BracketLeft
+                    {
+                        return Err("Failed to match left parenthesis.".to_string());
+                    }
+                    // Pop the left bracket.
+                    opt_stack.pop();
+                    // If there is unary operator before the left bracket, then pop it to the
+                    // postfix queue.
+                    if let Some(token) = opt_stack.last() {
+                        if let Btn::UnaryOpt(_) = token.btn {
+                            postfix.push(opt_stack.pop().unwrap());
+                        }
+                    }
+                }
+                _ => return Err(format!("Invalid token button {:?}", token.btn)),
+            };
+        }
+
+        // Pop any remaining operators from operator stack onto the postfix queue.
+        while opt_stack.last().is_some() {
+            postfix.push(opt_stack.pop().unwrap());
+        }
+
+        Ok(postfix)
+    }
+
+    /// Tokenize the `btn_stack`.
+    fn tokenize(&self) -> Vec<Token> {
+        let mut tokens: Vec<Token> = Vec::new();
+        for btn in &self.btn_stack {
+            let btn_expr = match btn.to_expr() {
+                Some(e) => e,
+                None => continue,
+            };
```

Comment:
Rust have a shorter way for this if you like, the `let else` syntax
```rs
let Some(btn_expr) = btn.to_expr() else {
    continue;
};
```

Reply:
Exactly what I was looking for